### PR TITLE
Feature/wcwpp 78 prepare logistics order

### PR DESCRIPTION
**Pull request by lucapircher**
## TLDR;

- Creates a logistics order for Gebüder Weiss from an WooCommerce Order
- Changes the customer id setting to a string, since it is required as a string in the API schema

## Additional Info

Please also validate that the mapping between the to objects has been implemented correctly.
You can find the documentation for the mapping  in [our Confluence](https://confluence.towa-digital.com/pages/viewpage.action?pageId=73372176). The documentation is a file at the bottom named `WooCommerce - OrderMapping.xlsx `. The relevant table in the Excel sheet is `LO Order Json Mapping`. Feel free to contact me if you have trouble finding it.

### CODE REVIEW

File: includes/LogisticsOrderFactory.php [hunk 1]
```diff
@@ -9,21 +9,151 @@
...
      * Creates a logistics order from a WooCommerce order
      *
      * @param object $wooCommerceOrder The order to be converted into a logistics order.
      * @return LogisticsOrder
      */
-    public static function buildFromWooCommerceOrder(object $wooCommerceOrder): LogisticsOrder
+    public function buildFromWooCommerceOrder(object $wooCommerceOrder): LogisticsOrder
+    {
+        $logisticsOrder = new LogisticsOrder();
+        $logisticsOrder->setUrl($this->settingsRepository->getSiteUrl() . "/wp-json/gebrueder-weiss-woocommerce/v1/update/" . $wooCommerceOrder->get_id());
```
Comment (philipp-hemmer) [inline includes/LogisticsOrderFactory.php:54]: we should shift the hardcoded part here into the .env since the callback-endpoint in OrderController is also hardcoded like this

File: includes/LogisticsOrderFactory.php [hunk 2]
```diff
@@ -9,21 +9,151 @@
...
+        $address->setName3($wooCommerceOrder->get_shipping_company());
+        $address->setStreet1($wooCommerceOrder->get_shipping_address_1());
+        $address->setStreet2($wooCommerceOrder->get_shipping_address_2());
+        $address->setCity($wooCommerceOrder->get_shipping_city());
+        $address->setZipCode($wooCommerceOrder->get_shipping_postcode());
+        $address->setCountryCode($wooCommerceOrder->get_shipping_country());
+        $address->setState($wooCommerceOrder->get_shipping_state());
+        $logisticsAddress->setAddress($address);
+
+        $fullName = implode(" ", array_filter([$wooCommerceOrder->get_shipping_first_name(), $wooCommerceOrder->get_shipping_last_name()]));
```
Comment (philipp-hemmer) [inline includes/LogisticsOrderFactory.php:93]: why d'you need array_filter here? implode concats array elements and the array is defined with the square brackets

File: includes/LogisticsOrderFactory.php [hunk 3]
```diff
@@ -9,21 +9,151 @@
...
+     * @return LogisticsAddress
+     */
+    private function createOrderByAddress(): LogisticsAddress
+    {
+        $logisticsAddress = new LogisticsAddress();
+        $logisticsAddress->setAddressType("orderby");
+
+        $addressReference = new AddressReference();
+        $addressReference->setQualifier("gwcustomerid");
+        $addressReference->setReference($this->settingsRepository->getCustomerId());
```
Comment (philipp-hemmer) [inline includes/LogisticsOrderFactory.php]: what does this addressReference do?

File: tests/Unit/LogisticsOrderFactoryTest.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+        $orderItemMethods = [
+            "get_id" => 123,
+            "get_product" => $product,
+            "get_quantity" => 4,
+        ];
+
+        /** @var MockInterface|stdClass */
+        $orderItem1 = Mockery::mock("WC_Order");
+        $orderItem1->allows($orderItemMethods);
```
Comment (philipp-hemmer) [inline tests/Unit/LogisticsOrderFactoryTest.php:54]: i'd prefer to have these methods in here to have a better overview

File: includes/LogisticsOrderFactory.php (hunk 2 above)
Comment (lucapircher) [reply includes/LogisticsOrderFactory.php:93]: The array filter removes null values from the array. If a person with a single name orders something, this prevents from a space beeing added at the end or beginning.

File: includes/LogisticsOrderFactory.php (hunk 1 above)
Comment (lucapircher) [reply includes/LogisticsOrderFactory.php:54]: Moving it to the .env file is the wrong move in my opinion, since the typical user of the plugin will not configure env variables. If we want to avoid this string, we could define it as a constant in the controller. However, I am fine with a string beeing used here.

File: includes/LogisticsOrderFactory.php (hunk 3 above)
Comment (lucapircher) [reply includes/LogisticsOrderFactory.php]: Good catch. I forgot to attach this one to the logistics address.

File: includes/LogisticsOrderFactory.php [hunk 5]
```diff
@@ -9,21 +9,152 @@
...
+        $address->setState($wooCommerceOrder->get_shipping_state());
+        $logisticsAddress->setAddress($address);
+
+        $fullName = implode(" ", array_filter([$wooCommerceOrder->get_shipping_first_name(), $wooCommerceOrder->get_shipping_last_name()]));
+
+        $contact = new Contact();
+        $contact->setName($fullName);
+        $contact->setEmail($wooCommerceOrder->get_billing_email());
+        $contact->setPhone($wooCommerceOrder->get_billing_phone());
+        $contact->setLanguage("de-De");
```
Comment (mattcrn) [inline includes/LogisticsOrderFactory.php]: "de-DE" according to excel

File: includes/LogisticsOrderFactory.php [hunk 6]
```diff
@@ -9,21 +9,152 @@
...
+        $logisticsAddress->setContact($contact);
+
+        return $logisticsAddress;
+    }
+
+    /**
+     * Creates the order by address from
```
Comment (mattcrn) [inline includes/LogisticsOrderFactory.php]: from?

File: includes/LogisticsOrderFactory.php [hunk 7]
```diff
@@ -9,21 +9,152 @@
...
+     *
+     * @return LogisticsAddress
+     */
+    private function createOrderByAddress(): LogisticsAddress
```
Comment (mattcrn) [inline includes/LogisticsOrderFactory.php]: this got me confused, create order by address, but it is actually create orderby address, so maybe createOrderbyAddress? 😄

File: includes/LogisticsOrderFactory.php (hunk 1 above)
Comment (mattcrn) [reply includes/LogisticsOrderFactory.php:54]: I thinks its fine here 👍

File: includes/LogisticsOrderFactory.php [hunk 8]
```diff
@@ -9,21 +9,152 @@
...
+        $logisticsOrder->setLogisticsAddresses([
+            $this->createConsigneeAddress($wooCommerceOrder),
+            $this->createOrderByAddress()
+        ]);
+
+        $logisticsOrder->setOrderLines(
+            $this->createOrderLines($wooCommerceOrder)
+        );
+
+        return $logisticsOrder;
```
Comment (mattcrn) [inline includes/LogisticsOrderFactory.php:67]: Is it possible, that the ownerId from the xls is missing?

**Review by mattcrn — CHANGES_REQUESTED** [review]

File: includes/LogisticsOrderFactory.php (hunk 7 above)
Comment (lucapircher) [reply includes/LogisticsOrderFactory.php]: Oh, now that I have read it again, it confused me too

**Review by mattcrn — APPROVED** [review]
